### ISSUE TITLE
Opam lock files do not work without `--unlock-base`

### ISSUE THREAD

**Issue by Lupus** [OCamlPro/opam-bin#25]
In CI I create opam switch as follows:

```
opam switch create . ocaml-base-compiler.4.08.1 --no-install
```

After that when I list installed packages, I get the following output:

```
# Packages matching: installed
# Name              # Installed                # Synopsis
base-bigarray       base+bin+e647d4c2+f2cc4984
base-threads        base+bin+f854a44a+f2cc4984
base-unix           base+bin+cf3e0c8e+f2cc4984
ocaml               4.08.1                     The OCaml compiler (virtual package)
ocaml-base-compiler 4.08.1                     Official release 4.08.1
ocaml-config        1                          OCaml Switch Configuration
```

Then I try to install dependencies with the following command:

```
opam install . --deps-only --locked --with-doc --with-test --working-dir
```

It complains about `base-threads` and suggests me to use `--unlock-base`. After adding `--unlock-base` I see the following during installation:

```
  - downgrade base-threads            base+bin+f854a44a+f2cc4984 to base
  - downgrade base-unix               base+bin+cf3e0c8e+f2cc4984 to base
```

`opam-bin.info` tells the following:

```
2020/10/03:21:03:32: base-bigarray.base+bin+e647d4c2+f2cc4984 already a binary package
2020/10/03:21:03:32: base-threads.base+bin+f854a44a+f2cc4984 already a binary package
2020/10/03:21:03:32: base-unix.base+bin+cf3e0c8e+f2cc4984 already a binary package
2020/10/03:21:03:38: ocaml-base-compiler.4.08.1 binary package installed from cache
2020/10/03:21:03:39: ocaml-config.1 binary package installed from cache
2020/10/03:21:03:39: ocaml.4.08.1 binary package installed from cache
2020/10/03:21:04:45: base-threads.base binary package installed from cache
2020/10/03:21:04:45: base-unix.base binary package installed from cache
```

Is this expected? I assume opam lock files should not include binary versions for those to work? It seems to work with the rest of the packages, which are listed as `binary package installed from cache` in `opam-bin.info`.

**Comment by lefessan** [comment]
I think the problem is that, when you install your compiler, `opam` selects binary packages for `base-threads` and `base-unix` (because they have a higher version, since `+bin...` is read as a version increment), they are then added to the "base" of the switch. Then, your lock file requires to use the `base` version, and complains that it's the `base+bin+...` version that is installed (and cannot be downgraded unless `--unlock-base` is provided). 

The solution would be to prevent `opam` from selecting binary packages for `base-threads` and `base-unix` in the first place.  The easiest way to achieve this while still using binary packages is to remove the binary package repository from the switch, with: `opam repo remove local-bin --set-default`
This way, binary packages will still be used using the cache mechanism of `opam-bin`, but they will never be selected by `opam` itself.

**Comment by lefessan** [comment]
Oh, I forgot that you are adding your own remote repository of binary packages. So, you should add it first, let `opam` download it, and then remove it with the upper command. Normally, it will remain in `~/.opam` and `opam-bin` will keep using it to find binary packages to replace source packages.

**Comment by Lupus** [comment]
> So, you should add it first, let opam download it, and then remove it with the upper command

Sounds like dark magic. Probably this needs to be at least documented. Maybe opam-bin could provide the users with some specialized command to add binary repository without adding it to opam?

I'll try this and let you know how it goes, thanks!

**Comment by lefessan** [comment]
Indeed, it's a good idea. Actually, `opam` has an option for that I think, `opam add repo bin-repo https://... --dont-select`. Then, you need to do `opam update --all` to sync it from time to time.

**Comment by Lupus** [comment]
That worked just fine. Probably the `--dont-select` option deserves mentioning in `opam-bin` documentation for such use case. Thanks for your help @lefessan !

**Comment by lefessan** [comment]
I changed the default behavior for the next version to use `--dont-select` when adding the `local-bin` local repository.

**Comment by Lupus** [comment]
May be it's worth adding configured http url as `remote-bin` with `--dont-select` automatically as well?

**Comment by lefessan** [comment]
You mean adding `remote-bin` when `base_url` is set in the configuration ?

**Comment by Lupus** [comment]
Yeah, this seems logical to me. User has some remote repo available via http, makes sense to add that to opam so that packages are available. My initial assumption was that `opam-bin` does that by default.

**Comment by lefessan** [comment]
If I remember well, the reason why I didn't added it is that you might want to create a new version of the binary repository without downloading the former binary archives. If the repo is added to `opam`, `opam-bin` will scan it to find binary archives.

**Comment by Lupus** [comment]
Well, adding the repository is easy, it probably just needs to be documented for folks who want to fetch packages from remote http repo as needed without doing `opam-bin pull`.
